### PR TITLE
CRM-8526 [ready for core team review] Added support for contribution tokens in emails sent via search results (Find contribution results and Advanced search when results are listed as contributions)

### PR DESCRIPTION
**Pull request by Edzelopez**
* [CRM-8526: Support \[contribution.custom_nn\] tokens in templates](https://issues.civicrm.org/jira/browse/CRM-8526)

### CODE REVIEW

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -5454,4 +5454,45 @@ public static function createProportionalFinancialEntries($entityParams, $lineIt
...
+   *
+   * @param string $html
+   *
+   * @param array $messageToken
+   *
+   * @param bool $escapeSmarty
+   *
+   * @return array
+   */
+  public static function replaceContributionTokens(
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: What function is used where elsewhere rendering these?

**Comment by totten** [comment]
I just want to give some high-level feedback - the "Send Email" screen is one of several things which sends email using the `CRM_Utils_Token` pattern (e.g. repeatedly calling `CRM_Utils_Token::replaceFoo()`). During the lead-up to v4.7.0, I worked on refactoring "Scheduled Reminders" to support new entities and their tokens. The whys/hows are discussed more in:

https://wiki.civicrm.org/confluence/display/CRMDOC/Token+Reference

The original work for v4.7.0 introduced `TokenProcessor` but only adopted it for "Scheduled Reminders". The ambition was to use it for others, but now-a-days that ambition needs to be channeled through LExIM. Example: instead of making a whole bunch of patches to `CRM/Mailing/BAO/Mailing{,Job}.php`,
I made a "leap" to replace the CiviMail backend with an extension ([flexmailer](https://github.com/civicrm/org.civicrm.flexmailer/)) . Among other things, the extension uses `TokenProcessor` for mail-blasts.

CRM-8526 seems to touch squarely on one of the problems that motivated the `CRM_Utils_Token` =>`TokenProcessor` switch -- it's hard to insert new, context-aware tokens. ("Context-ware" -- for example, if you send an email about a contribution, then the `contribution_id` is part of the context, and it gives access to a whole bunch of tokens. Or if you send an email about an event, then the `event_id` and `participant_id` could be part of the context, giving access to several other tokens.)

My gut says that some kind of leap might be in order for the 1-on-1 mailings, but that might be a big thing, and I'm afraid that I don't have a lot of time to think about how. But I wanted to share the link and put this on your radar.

File: CRM/Contribute/BAO/Contribution.php (hunk 1 above)
Comment (Edzelopez) [inline CRM/Contribute/BAO/Contribution.php]: @eileenmcnaughton Sorry, I didn't get what you meant to say. Could you please explain a bit more?

**Comment by Edzelopez** [comment]
Thanks @totten 

I actually did have a look before implementing the fix in the current way. It seemed from the scheduled reminders code that the token processor made use of the $dao object which was constructed using the actionscheduler query. In my case, I didn't have the $dao object since it was a search based form, so I assumed that support for this area was yet to be done. I agree that some refactoring is needed in 1-on-1 mailings for all components, but with the current code, it was too huge of a refactor.

**Comment by Edzelopez** [comment]
@monishdeb did you have a chance to QA/check if this can be extended using @totten's suggestion?

File: CRM/Contribute/BAO/Contribution.php (hunk 1 above)
Comment (JoeMurray) [inline CRM/Contribute/BAO/Contribution.php]: @Edzelopez : here's what I think @eileenmcnaughton meant: elsewhere in code when rendering these contribution tokens (assuming that is supported somewhere), what function does the replacement of the tokens with their values.

**Comment by JoeMurray** [comment]
Monish provides a rough estimate of 6 - 8 days of work to reimplement this functionality using Tim's approach. We don't have funding for that so we propose either reverting the merge done by Jitendra or merging the fix provided here by Edsel.

**Comment by colemanw** [comment]
@JoeMurray I think this is the one we discussed where you are providing a bugfix and wanting to avoid scope creep for now.
@eileenmcnaughton and I discussed and are ok with this going in. @Edzelopez can you please rebase it?

**Comment by monishdeb** [comment]
@colemanw @Edzelopez I have rebased the PR. However, there is a test build failure reflecting on every other PR and I am afraid this will also reflect in current test build.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I think the test fail might be the DAO::testObject failing to identify the DAO for SMSProvider

**Pull request merged**